core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java

```diff
@@ -170,6 +170,13 @@ private DataBuffer readChunkInternal() throws IOException {
       return null;
     }
     mPosToRead += buffer.readableBytes();
+    if (mPosToRead - mReadRequest.getOffset() == mReadRequest.getLength()) {
+      // finish reading all the data, close gRPC stream earlier to release resources
+      // potentially improve machine learning millions of small file read throughput
+      LOG.debug("Finished reading all data, close gRPC stream");
+      close();
```

Comment:
@ggezer @LuQQiu 

1 buffer here is built based upon `response.getChunk().getData().asReadOnlyByteBuffer()`.
if we close the stream before returning the buffer, is the reference to this buffer still safe?

2 what if `close` throws an exception, do we need to release any resource related to buffer?

Reply:
The 1 worry is about whether gRPC stream close() will release the Chunk in ReadResponse. I don't think it will, but better to be confirmed by @ggezer. I didn't see any code or doc indicate that scenario.

Release buffer when close throw an exception.
Considering still return the byteBuffer to client if close throws exception, but in that case, do we just rethrow that exception when the actual close() is called by the client

Reply:
Currently, grpc data reader offloads releasing to the caller. By moving proactive closing logic to BlockInStream level, you wouldn't have to deal with this as well.